File: src/cont_move_pi.cpp
#include <ros/ros.h>
#include <std_msgs/String.h>
#include <pigpio.h>
#include <string>
#include <iostream>
#include <chrono>

#define PUL_PIN 24
#define DIR_PIN 23
#define STEPS_PER_REV 400

void step() {
    gpioWrite(PUL_PIN, 1);
    ros::Duration(0.00025).sleep(); // 100ms delay
    gpioWrite(PUL_PIN, 0);
    ros::Duration(0.00025).sleep(); // 100ms delay
}

void continuous_rotation(int rev_per_minute) {
    double delay = 0.00001; // 0.001 seconds delay
    ROS_INFO("Rotating motor at %d RPM with %f seconds delay per step", rev_per_minute, delay);
    for (int i = 0; i < STEPS_PER_REV; ++i) { // One full revolution
        step();
    }
}

void command_callback(const std_msgs::String::ConstPtr& msg) {
    std::string command = msg->data;
    std::transform(command.begin(), command.end(), command.begin(), ::tolower);
    ROS_INFO("Received command: %s", command.c_str());
    
    if (command == "step") {
        step();
        ROS_INFO("Executed one step");
    } else if (command == "rotate") {
        gpioWrite(DIR_PIN, 1); // Ensure direction is set before rotating
        continuous_rotation(100); // Rotate at 100 RPM
        ROS_INFO("Executed continuous rotation");
    } else if (command == "stop") {
        ROS_INFO("Stopping...");
        ros::shutdown();
    }
}

int main(int argc, char** argv) {
//    if (gpioInitialise() < 0) {
//        std::cerr << "Failed to initialize GPIO." << std::endl;
//        return 1;
//    }
    
    gpioSetMode(PUL_PIN, PI_OUTPUT);
    gpioSetMode(DIR_PIN, PI_OUTPUT);
    gpioWrite(DIR_PIN, 0);
    gpioWrite(PUL_PIN, 0); // Ensure PUL_PIN starts LOW
    
    ros::init(argc, argv, "motor_command_subscriber");
    ros::NodeHandle nh;
    ros::Subscriber sub = nh.subscribe("motor_commands", 10, command_callback);
    
    ROS_INFO("Motor controller node started.");
    ros::spin();
    
    gpioTerminate();
    return 0;
}

File: scripts/tests/motors/cont_move_pi.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import String
import RPi.GPIO as GPIO
import time

# GPIO setup
PUL_PIN = 18
DIR_PIN = 16

GPIO.setmode(GPIO.BOARD)
GPIO.setup(PUL_PIN, GPIO.OUT)
GPIO.setup(DIR_PIN, GPIO.OUT)

GPIO.output(DIR_PIN, GPIO.LOW)

def step():
    GPIO.output(PUL_PIN, GPIO.HIGH)
    time.sleep(0.0005)  # smallest delay possible
    GPIO.output(PUL_PIN, GPIO.LOW)
    time.sleep(0.0005)  # smallest delay possible

def continuous_rotation(rev_per_minute):
    delay = 60.0 / (400 * rev_per_minute)  # 400 steps per revolution
    for _ in range(400):  # One full revolution
        step()

def command_callback(msg):
    command = msg.data.lower()
    rospy.loginfo(f"Received command: {command}")
    
    if command == "step":
        step()
        rospy.loginfo("Executed one step")
    elif command == "rotate":
        continuous_rotation(100)  # Rotate at 100 RPM
        rospy.loginfo("Executed continuous rotation")
    elif command == "stop":
        rospy.loginfo("Stopping...")
        rospy.signal_shutdown("Stop command received")

def motor_control_listener():
    rospy.init_node('motor_command_subscriber', anonymous=True)
    rospy.Subscriber('motor_commands', String, command_callback)
    rospy.spin()

if __name__ == '__main__':
    try:
        motor_control_listener()
    except rospy.ROSInterruptException:
        pass
    finally:
        GPIO.cleanup()

